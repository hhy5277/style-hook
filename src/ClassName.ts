/** convert the charCode to [0-9A-Za-z] */
function toWord(charCode: number) {
  // 0-9 : 48-57
  // A-Z : 65-90
  // a-z : 97-122
  const scope = 10 + 26 + 26
  charCode = charCode % scope
  if(charCode < 10)
    return String.fromCharCode(48 + charCode)
  charCode -= 10
  if(charCode < 26)
    return String.fromCharCode(65 + charCode)
  charCode -= 26
  return String.fromCharCode(97 + charCode)
}

const ICON_EMOJI = 'ðŸŽ¨'  // to avoid className start with number

/** compute the hash of the code simply */
export default function (code: string, length = 8) {
  const classNameCodes: number[] = new Array(length).fill(code.length)
  for(let i=0; i<code.length; i++)
    classNameCodes[i%length] += code.charCodeAt(i)
  return ICON_EMOJI + classNameCodes.map(toWord).join('')
}
