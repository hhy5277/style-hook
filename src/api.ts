export * from './hooks'
export * from './media'
export { default as css } from './css'
export { Palette } from './palette'

const StyleHook = {}
export default StyleHook
