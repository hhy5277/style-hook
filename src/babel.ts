export function cssMinify(code: string) {
  const trimWith = (s: string) => RegExp(`\\s*\\${s}\\s*`, 'g')
  const trimStartWith = (s: string) => RegExp(`\\s+\\${s}`, 'g')
  const trimEndWith = (s: string) => RegExp(`\\${s}\\s+`, 'g')
  return code
    .trim()

    // remove comments
    .replace(/\/\*.*?\*\//g, '')

    // remove unnecessary spaces
    .replace(/\s{2,}/g, ' ')
    .replace(trimWith(':'), ':')
    .replace(trimWith(';'), ';')
    .replace(trimWith(','), ',')
    .replace(trimEndWith('('), '(')
    .replace(trimStartWith(')'), ')')
    .replace(trimWith('{'), '{')
    .replace(trimWith('}'), '}')

    // remove unnecessary semicolons
    .replace(/;;+/, ';')
    .replace(/;(?=\})/, '')
    .replace(/;$/, '')

    // shorter number
    .replace(/\b0\.(\d+)/, '.$1')
    .replace(/\b0[a-zA-z]+/, '0')
}


type Babel = typeof import('@babel/core')

const apiName = [
  'useStyle',
  'useModuleStyle',
  'useGlobalStyle',
  'createStyleModule',
  'css',
]

export default function babelPluginStyleHook(babel: Babel): babel.PluginObj {
  const t = babel.types
  return {
    visitor: {
      TaggedTemplateExpression(path) {
        const { tag, quasi } = path.node
        if (t.isIdentifier(tag) && apiName.includes(tag.name)) {
          const quasiPath = path.get('quasi')
          const { quasis, expressions } = quasi
          let expressionSymbol = 'VAR'
          while(quasis.map(tplEl => tplEl.value.raw).join(' ').includes(expressionSymbol))
            expressionSymbol += '_'
          const sourceCode = quasis.map((tplEl, i) => (
            tplEl.value.raw
            + (i !== quasis.length -1 ? `${expressionSymbol}${i}` : '')
          )).join('')
          const minCode = cssMinify(sourceCode)
          quasiPath.replaceWith(
            t.templateLiteral(
              minCode.split(RegExp(`${expressionSymbol}\\d+`)).map(raw => (
                t.templateElement({ raw, cooked: raw }, true)
              )),
              expressions.filter((_, i) => RegExp(`${expressionSymbol}${i}`).test(minCode))
            )
          )
        }
      }
    }
  }
}
