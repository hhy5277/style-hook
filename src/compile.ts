/** compile the source code to css */
export default function(code: string, selector: string) {
  const codes = code.trim()
    .replace(/\s*([;{}])\s*/g, '$1\n')
    .split(/\n/)
    .filter(x => x)
  type Styles = {[selector: string]: string}
  const mediaStyles = {} as {[meida: string]: Styles}
  const selectorStack: string[] = [] // not include currentSelector
  let currentSelector = selector
  let currentMedia = ''
  let hideScoped = false
  const getCurrentStyles = () => {
    if(!mediaStyles[currentMedia])
      mediaStyles[currentMedia] = {}
    return mediaStyles[currentMedia]
  }
  const toChil = (chilSelector: string): void => {
    // TODO: split with ","
    if(chilSelector.startsWith('@media')) {
      currentMedia = chilSelector
      if(!mediaStyles[currentMedia])
        mediaStyles[currentMedia] = {}
      return
    }
    selectorStack.push(currentSelector)
    if(chilSelector.startsWith('false')) {
      hideScoped = true
      currentSelector = 'false'
      return
    }
    chilSelector = chilSelector.replace(/^true\s*/, '')
    if(chilSelector.startsWith('&'))
      currentSelector = currentSelector + chilSelector.slice(1)
    else if(chilSelector.endsWith('&'))
      currentSelector = chilSelector.slice(0, -1) + currentSelector
    else if(chilSelector.startsWith(':'))
      currentSelector = currentSelector + chilSelector
    else
      currentSelector = [currentSelector, chilSelector].join(' ').trim()
  }
  const toParent = (code: string): void => {
    if(code.length > 1)
      addCodeToCurrentScope(code.replace(/;?\s*\}/, ';'))
    if(hideScoped && selectorStack.every(selector => selector !== 'false'))
      hideScoped = false
    if(currentMedia) return void(currentMedia = '')
    const parentSelector = selectorStack.pop()
    if(!parentSelector) throw 'selectorStack error'
    currentSelector = parentSelector
  }
  const addCodeToCurrentScope = (code: string) => {
    if(hideScoped) return
    let styleSheets = getCurrentStyles()
    if(!styleSheets[currentSelector])
      styleSheets[currentSelector] = ''
    // TODO: rpx prefix
    styleSheets[currentSelector] += code
  }
  for(let codeLine of codes) {
    if(codeLine.endsWith(';'))
      addCodeToCurrentScope(codeLine)
    else if(codeLine.endsWith('{'))
      toChil(codeLine.slice(0, -1).trim())
    else if(codeLine.endsWith('}'))
      toParent(codeLine)
    else
      addCodeToCurrentScope(codeLine + ';')
  }
  return Object.entries(mediaStyles).map(([media, styles]) => {
    const code = Object.entries(styles)
      .map(([selector, code]) => `${selector}{${code}}`)
      .join('\n')
    if(media) return `${media}{\n${code}\n}`
    return code
  }).join('\n')
}
