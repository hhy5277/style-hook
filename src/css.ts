export default function (template: TemplateStringsArray, ...substitutions: any[]) {
  substitutions = substitutions.concat('')
  const filterEmpty = (s: any) =>
    [null, undefined, false].includes(s)
      ? '' : s
  const mapTpl = (tpl: string, i: number) =>
    tpl + filterEmpty(substitutions[i])
  return template
    .map(mapTpl)
    .join('')
}
