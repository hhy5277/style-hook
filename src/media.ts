import React, { useEffect, useState } from 'react'

interface MediaQueryProps {
  breakPoint?: 'XS' | 'S' | 'M' | 'L' | 'XL' | 'XXL'
  minWidth?: number
  maxWidth?: number
  minHeight?: number
  maxHeight?: number
}
function mediaQuery(query: MediaQueryProps): boolean {
  if (query.breakPoint)
    return mediaQuery(MediaQuery[query.breakPoint])
  if (typeof query.maxWidth === 'number' && innerWidth > query.maxWidth)
    return false
  if (typeof query.minWidth === 'number' && innerWidth < query.minWidth)
    return false
  if (typeof query.maxHeight === 'number' && innerHeight > query.maxHeight)
    return false
  if (typeof query.minHeight === 'number' && innerHeight < query.minHeight)
    return false
  return true
}

export function useMediaQuery(query: MediaQueryProps) {
  const [match, setMatch] = useState(() => mediaQuery(query))
  useEffect(() => {
    const handleResize = () => setMatch(mediaQuery(query))
    window.addEventListener('resize', handleResize)
    return () => window.removeEventListener('resize', handleResize)
  }, [])
  return match
}

/**
 * @example <MediaQuery breakPoint="XS">you can see me in the phone.</MediaQuery>
 */
export function MediaQuery(props: React.PropsWithChildren<MediaQueryProps>) {
  const match = useMediaQuery(props)
  return match
    ? React.createElement(React.Fragment, {}, props.children )
    : null
}


type ScreenSize = 'XS' | 'S' | 'M' | 'L' | 'XL' | 'XXL'
class Screen {
  static sizes = ['XS', 'S', 'M', 'L', 'XL', 'XXL']
  static breakPoints = [576, 768, 992, 1200, 1920]
  size!: ScreenSize
  minWidth = 0
  maxWidth = Infinity
  constructor(size: ScreenSize) {
    const i = Screen.sizes.indexOf(size)
    this.size = size
    if (i !== -1) {
      if (size !== 'XS')
        this.minWidth = Screen.breakPoints[i - 1]
      if (size !== 'XXL')
        this.maxWidth = Screen.breakPoints[i] - 1
    }
  }
  toString() {
    const bebabCase = (name: string) => name.replace(/[A-Z]/g, $ => `-${$.toLowerCase()}`)
    return `@media screen ${
      (<('minWidth'|'maxWidth')[]>['minWidth', 'maxWidth'])
        .filter(property => ![0, Infinity].includes(this[property]))
        .map(property => `and (${bebabCase(property)}:${this[property]}px)`)
        .join(' ')
    }`
  }
}

/** Extra small devices (portrait phones, less than 576px) */
MediaQuery.XS = new Screen('XS')
/** Small devices (landscape phones, 576px and up, but less than 768px) */
MediaQuery.S = new Screen('S')
/** Medium devices (tablets, 768px and up, but less than 992px) */
MediaQuery.M = new Screen('M')
/** Large devices (desktops, 992px and up, but less than 1200px) */
MediaQuery.L = new Screen('L')
/** Extra large devices (large desktops, 1200px and up, but less than 1920px) */
MediaQuery.XL = new Screen('XL')
/** Very very large devices (1K/2K/... desktops, 1920px and up) */
MediaQuery. XXL = new Screen('XXL')

/** select the range between the screen */
MediaQuery.range = function (a: ScreenSize, b?: ScreenSize) {
  if (typeof b === 'undefined')
    return MediaQuery[a].toString()
  const minWidth = Math.min(
    MediaQuery[a].minWidth,
    MediaQuery[b].minWidth,
  )
  const maxWidth = Math.max(
    MediaQuery[a].maxWidth,
    MediaQuery[b].maxWidth,
  )
  return Screen.prototype.toString.call({
    minWidth, maxWidth
  })
}
