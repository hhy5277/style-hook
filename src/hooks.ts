import { useLayoutEffect, useMemo, useEffect, useState } from 'react'
import compile from './compile'
import ClassName from './ClassName'
import StyleSheetsManager from './StyleSheetsManager'
import css from './css'
import { getCSSModule, isSSR } from './utils'

function useCss(template: TemplateStringsArray, ...substitutions: any[]) {
  const sourceCode = css(template, ...substitutions)
  const className = useMemo(() => ClassName(sourceCode), substitutions)
  return { sourceCode, className }
}

function useStyleCode(id: string, getCode: () => string) {
  if (isSSR) return
  useLayoutEffect(() => {
    StyleSheetsManager.useStyle(id, getCode)
    return () => StyleSheetsManager.unUseStyle(id)
  }, [id])
}

function useInnerWidth() {
  if (isSSR) return NaN
  const [width, setWidth] = useState(window.innerWidth)
  useEffect(() => {
    const updateWidth = () => setWidth(window.innerWidth)
    window.addEventListener('resize', updateWidth)
    return () => window.removeEventListener('resize', updateWidth)
  }, [])
  return width
}

export function useStyle(template: TemplateStringsArray, ...substitutions: any[]) {
  const { sourceCode, className } = useCss(template, ...substitutions)
  const innerWidth = useInnerWidth()
  useStyleCode(`${className}[#${innerWidth}]`, () => (
    compile(sourceCode, `.${className}`)
  ))
  return className
}

export function useGlobalStyle(template: TemplateStringsArray, ...substitutions: any[]) {
  const { sourceCode, className: hash } = useCss(template, ...substitutions)
  const innerWidth = useInnerWidth()
  useStyleCode(`${hash}[global][#${innerWidth}]`, () => (
    compile(sourceCode, '').replace(/(^|\})\{/, '$1html{')
  ))
}

export interface CSSModule {
  [className: string]: string
}
export function useModuleStyle(template: TemplateStringsArray, ...substitutions: any[]): CSSModule {
  const { sourceCode, className: moduleName } = useCss(template, ...substitutions)
  const innerWidth = useInnerWidth()
  const { styles, sourceCodeWithModuleId } = useMemo(() => {
    return getCSSModule(sourceCode, moduleName)
  }, [moduleName])
  useStyleCode(`${moduleName}[module][#${innerWidth}]`, () => (
    compile(sourceCodeWithModuleId, '')
  ))
  return styles
}
