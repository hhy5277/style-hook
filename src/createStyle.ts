import css from './css'
import { CSSModule } from './hooks'
import compile from './compile'
import StyleSheetsManager from './StyleSheetsManager'
import { getCSSModule, isSSR } from './utils'

export function createModuleStyle(template: TemplateStringsArray, ...substitutions: any[]): CSSModule {
  let sourceCode = css(template, ...substitutions)
  if (isSSR) return getCSSModule(sourceCode).styles
  const cssModule: CSSModule = {}
  let cssCode = ''
  const updateStyle = () => {
    if (cssCode) StyleSheetsManager.removeStyle(cssCode)
    const { styles, sourceCodeWithModuleId } = getCSSModule(sourceCode)
    Object.assign(cssModule, styles)
    cssCode = compile(sourceCodeWithModuleId, '')
    StyleSheetsManager.addStyle(cssCode)
  }
  updateStyle()
  window.addEventListener('resize', updateStyle)
  return cssModule
}
