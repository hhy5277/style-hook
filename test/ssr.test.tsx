/**
 * @jest-environment node
 */

import React from 'react'
import ReactDOMServer from 'react-dom/server'
import { createModuleStyle, useGlobalStyle } from '../src/api'


describe('ssr', () => {
  test('render hook correct without jsdom', () => {
    function Test() {
      useGlobalStyle`color: red;`
      return <div>ok</div>
    }
    const html = ReactDOMServer.renderToString(<Test />)
    expect(html).toMatchSnapshot()
  })
  test('render createStyle correct without jsdom', () => {
    const styles = createModuleStyle ` .a {} .b {} .c {} `
    expect(Object.keys(styles)).toEqual(['a', 'b', 'c'])
  })
})
