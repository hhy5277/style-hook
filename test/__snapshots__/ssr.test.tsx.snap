// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr render hook correct without jsdom 1`] = `"<div data-reactroot=\\"\\">ok</div>"`;
